LFS-535: Solve concurrency issues when uploading somatic variant files as CSVs in bulk

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-535).

Allows the upload of multiple files at once, or one at a time with multiples selected.

![image](https://user-images.githubusercontent.com/4656440/117047998-804b4380-ace0-11eb-8107-1ba004faa544.png)

![image](https://user-images.githubusercontent.com/4656440/117047975-775a7200-ace0-11eb-9671-e6b1a0e1daa4.png)

**To test:** no combination of switching subject/tumor names, uploading single files vs uploading all, or clicking things out of order should break anything.

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Allow users to upload files one at a time, and select multiple files at
once

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Allow the upload of multiple somatic variant files at once.

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Fix the other files display when switching the name of the
subject/tumor/regionname.

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Fix a graphical glitch where previously uploaded files would appear to
upload twice when uploading via the upload One followed by an uploadAll
call.

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Right align the progress bar and prevent a bug that was causing the
bars to always appear full.

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Fix various docstrings

sdumitriu commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -205,11 +211,30 @@ export default function VariantFilesContainer() {
+      return fetchWithReLogin(globalLoginDisplay, url)
+        .then((response) => response.ok ? response.json() : Promise.reject(response))
+        .then((json) => {
+          processedFile.sameFiles = json.rows.filter((row) => row["@name"] === fileName);
```
I wonder if we should check the file name at all, or just the subject... @marta- @k-pace what is the intended behavior:

1. other somatic variant files exist for this patient / tumor / region, regardless of their name, or
2. the same file was uploaded for this patient / tumor / region?

In most cases, these two will be the same, since we derive the subject from the filename, so it's unlikely to have files with different names for the same subject.

marta- replied:
The text to the user reads "There are {N} other versions of this file" so checking how many times we have a file with this name in the database makes sense.

Once the user adjusts the subject ids (if applicable) and uploads the file, it may be useful to inform them of other variant files that were uploaded to that subject, but that's out of scope here. Please report as a separate task for discussion.

sdumitriu commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -626,6 +696,13 @@ export default function VariantFilesContainer() {
+  /**
+   * Generate the JSON representation of a single subject
+   *
+   * @param {string} refType The jcr:reference:type for this subject (e.g. "Patient")
```
```suggestion
   * @param {string} refType The type for this subject, without the `/SubjectType/` path prefix, e.g. "Patient/Tumor"
```

OrdiNeu replied:
Fixed in `96d498c` :heavy_check_mark:

marta- committed:
LFS-535: Solve concurrency issues when uploading somatic variant files as CSVs in bulk
Various UI improvements:
* Fix left alignments (drop zone, file info and "Upload all" button all had different indentations)
* Only show "Upload all" if there's more than one file to upload
* Improve the look of the progress bar: use LinearProgress instead of custom elements, display it under the file name instead of floating to the right

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -808,16 +881,24 @@ export default function VariantFilesContainer() {
-      }
+      { selectedFiles?.length > 1 ?
+      <Grid item>
+      <Button type="submit" variant="contained" color="primary" disabled={uploadInProgress || !!error && selectedFiles.length == 0} form="variantForm">
```
Can we hide this button when all files are finished uploading?

OrdiNeu replied:
Sure! Applied in `9cc2843` :heavy_check_mark:

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Apply docstring fix suggested by @sdumitriu

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Hide the Upload All button if there are no unsent files.

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Hide the upload all button if only one file is selected.

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Fix the display of the upload all button to disable instead of hiding
when all files are uploaded.

sashaandjic commented:
Can we disable UPLOAD ALL bttn once all files are uploaded?
![image](https://user-images.githubusercontent.com/23243244/117203188-bbbb4000-adbc-11eb-976e-b1048714ef4d.png)

sashaandjic commented:
Selectin this (4 files, 2 with incorrect name format):
![image](https://user-images.githubusercontent.com/23243244/117203830-806d4100-adbd-11eb-9e24-6504db5293a3.png)

Would result in this:
![image](https://user-images.githubusercontent.com/23243244/117203861-895e1280-adbd-11eb-8e9a-04abab3b5690.png)

I**t would be nice if the other file with the wrong name  format was mentioned.**

However, selecting this ( first file with incorrect format, and 4 with correct):
![image](https://user-images.githubusercontent.com/23243244/117204078-d17d3500-adbd-11eb-8f5c-f85ae727fcfa.png)

Results in this:
![image](https://user-images.githubusercontent.com/23243244/117204098-d9d57000-adbd-11eb-8276-25c9d3019997.png)

marta- commented:
> Can we disable UPLOAD ALL bttn once all files are uploaded?

@sashaandjic This has already been done in the last commit, you might not have the latest code.

marta- commented:
> I**t would be nice if the other file with the wrong name format was mentioned.**

Agreed. Looks like the way we set and display the error needs some rework.

sashaandjic commented:
The last commit introduced the following problem. I was trying to upload 3 files: 2 with proper format and 1 with a wrong one. Because one of them had a wrong format, I can no longer upload the other two.
![image](https://user-images.githubusercontent.com/23243244/117217414-57ee4280-adcf-11eb-8dae-43b2eacce812.png)

sashaandjic commented:
I had no issues uploading multiple files of a different format at the same time. I was also able to rename subjects for some of them without any issue.
The problems reported above can be logged separately.

marta- commented:
> The last commit introduced the following issue. I was trying to upload 3 files: 2 with proper format and 1 with a wrong one. Because one of them had a wrong format, the UPLOAD ALL was disabled. I can upload each one separately.

I confirm. @OrdiNeu I don't see we need to disable the "upload all" button on error, but maybe I'm missing something.

sashaandjic reviewed (approved):
tested

marta- reviewed (changes requested):
Let's fix the error reporting and disabled "Upload all", per @sashaandjic 's comments.

sashaandjic commented:
How should we treat this situation?
In the Variant Upload form I upload the same file 2 times.  
The Variant Upload knows and displays the message stating that there is another version of the file.
Then I go to the Somatic Variants form and upload the same file one more time.
Should Variant Upload form be aware of the third version? Right now it isn't.

marta- commented:
> How should we treat this situation?
> In the Variant Upload form I upload the same file 2 times.
> The Variant Upload knows and displays the message stating that there is another version of the file.
> Then I go to the Somatic Variants form and upload the same file one more time.
> Should Variant Upload form be aware of the third version? Right now it isn't.

The "versions" that the variant uploader is reporting are files with the same name (not overwritten by a subsequent upload) currently in use anywhere in the database. TBH I find this confusing myself, and "versions of this file" is definitely not the best wording for it as it can be misinterpreted. It should be reworked as a separate task together with the question that @sdumitriu raised above about what files we should be counted (and what message should be displayed).

OrdiNeu committed:
LFS-535: Solve concurrency issues when uploading somatic variant files
as CSVs in bulk

Fix an issue where an invalid filename would stop processing and only
display one error.
Fix an issue where the upload buttons were incorrectly disabled during
error status.

OrdiNeu commented:
@sashaandjic 
> However, selecting this ( first file with incorrect format, and 4 with correct): 

This should be fixed in `cfe0a0c` :heavy_check_mark: 

@sashaandjic 
> The last commit introduced the following issue. I was trying to upload 3 files: 2 with proper format and 1 with a wrong one. Because one of them had a wrong format, the UPLOAD ALL was disabled. I can upload each one separately.

This should be fixed in `cfe0a0c` :heavy_check_mark:

sashaandjic reviewed (approved):
tested
